### ISSUE TITLE
suspected memory leak of qpair free in nvme_pcie_ctrlr_delete_io_qpair

### ISSUE THREAD

**Issue by malasong227** [spdk/spdk#716]
## Expected Behavior
No memory resource leaking

## Current Behavior
qpair memory leak when delete io qpair fail

nvme_pcie_ctrlr_delete_io_qpair(struct spdk_nvme_ctrlr *ctrlr, struct spdk_nvme_qpair *qpair)
{
	struct nvme_completion_poll_status status;
	int rc;

	assert(ctrlr != NULL);

	if (ctrlr->is_removed) {
		goto free;
	}

	/* Delete the I/O submission queue */
	status.done = false;
	rc = nvme_pcie_ctrlr_cmd_delete_io_sq(ctrlr, qpair, nvme_completion_poll_cb, &status);
	if (rc != 0) {
		return rc;
	}
	while (status.done == false) {
		spdk_nvme_qpair_process_completions(ctrlr->adminq, 0);
	}
	if (spdk_nvme_cpl_is_error(&status.cpl)) {
		return -1;
	}

when nvme_pcie_ctrlr_cmd_delete_io_sq or spdk_nvme_wait_for_completion fails, nvme_pcie_ctrlr_delete_io_qpair will return. But before it return, we not execute 
nvme_pcie_qpair_destroy(qpair) to free qpair memory.

Is there qpair memory leak when delete sq or cq fails.

## Possible Solution
add nvme_pcie_qpair_destroy(qpair) before return rc and retrun -1

	/* Delete the I/O submission queue */
	rc = nvme_pcie_ctrlr_cmd_delete_io_sq(ctrlr, qpair, nvme_completion_poll_cb, &status);
	if (rc != 0) {
+		nvme_pcie_qpair_destroy(qpair);
		return rc;
	}
	if (spdk_nvme_wait_for_completion(ctrlr->adminq, &status)) {
	    SPDK_ERRLOG("nvme_delete_io_sq failed! sc[0x%x], sct[0x%x]\n", status.cpl.status.sc, status.cpl.status.sct);
+		nvme_pcie_qpair_destroy(qpair);
		return -1;
	}

**Comment by malasong227** [comment]
@ZiyeYang @jimharris

**Comment by optimistyzy** [comment]
@malasong227 If the operation is failed, there may still be outstanding requests in the cq list. So if you directly call the destroy function,  will cause some issues. So I think that the better solution is to track the status and return, and let the upper layer to do the destroy again instead of doing the bruteforce destruction here. This is my understanding.   I think that @jimharris  will provide more useful thoughts.

**Comment by jimharris** [comment]
I agree with @optimistyzy that we should not directly destroy the qpair when one of these functions fail.  But we also do not have tests to show that user can actually recover in one of these cases.  These functions can only fail if:

1) we have no more request objects (128 are allocated currently) to issue the delete_sq or delete_cq command
2) the delete_sq or delete_cq command failed at the controller

@optimistyzy - can you please look into the error handling on this case further?

**Comment by optimistyzy** [comment]
@jimharris  For 1, only if the request cannot be allocate due to nomem issue, otherwise, it will be queued into the submission list, it will be finally sent to the drive.

**Comment by jimharris** [comment]
@optimistyzy - if a request cannot be allocated, it cannot be queued.  You may be thinking of the nvme_trackers - if there are no nvme_trackers available, than an nvme_request cannot be submitted to the driver.  It will be queued until an nvme_tracker becomes available after a completion.

**Comment by optimistyzy** [comment]
@jimharris  For the case you mentioned, the return value (rc =0),  so it will not cause the return value (rc !=0) case.  I would like to add some error message first in this function nvme_pcie_ctrlr_delete_io_qpair
